### PR TITLE
Add type stubs for static type checking

### PR DESCRIPTION
**Pull request by sbdchd**
Type stubs for the C extension and the necessary packaging config so
type checkers will find the stubs.

fixes https://github.com/redis/hiredis-py/issues/99

rel: https://www.python.org/dev/peps/pep-0561/#packaging-type-information

### CODE REVIEW

File: hiredis/hiredis.pyi [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        encoding: Optional[str] = ...,
+        errors: Optional[str] = ...,
+    ) -> None: ...
+    def feed(
+        self, __buf: Union[str, bytes], __off: int = ..., __len: int = ...
+    ) -> None: ...
+    def gets(self, __shouldDecode: bool = ...) -> Any: ...
+    def setmaxbuf(self, __maxbuf: Optional[int]) -> None: ...
+    def getmaxbuf(self) -> int: ...
+    def len(self) -> int: ...
```
Comment (sbdchd) [inline hiredis/hiredis.pyi:21]: future PR, but might be useful to setup `mypy` or one of the other type checkers to help catch any potential type issues by type checking the `test` module

File: hiredis/hiredis.pyi [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        self,
+        protocolError: Callable[[str], Exception] = ...,
+        replyError: Callable[[str], Exception] = ...,
+        encoding: Optional[str] = ...,
+        errors: Optional[str] = ...,
+    ) -> None: ...
+    def feed(
+        self, __buf: Union[str, bytes], __off: int = ..., __len: int = ...
+    ) -> None: ...
+    def gets(self, __shouldDecode: bool = ...) -> Any: ...
```
Comment (sbdchd) [inline hiredis/hiredis.pyi:18]: might be able to make this return type more precise, but that might also cause more issues for consumers of the library

File: hiredis/hiredis.pyi [hunk 3]
```diff
@@ -0,0 +1,25 @@
+from typing import Any, Callable, Optional, Union
```
Comment (sbdchd) [inline hiredis/hiredis.pyi:1]: I checked these types by looking at the c source code and running mypy against the `test` module with the following setup:


```
; mypy.ini
[mypy]
show_column_numbers=True
pretty=False
show_error_codes = True

disallow_any_unimported=True
disallow_any_expr=False
disallow_any_decorated=True
disallow_any_explicit=False
disallow_any_generics=True
disallow_subclassing_any=True

disallow_untyped_calls=True
disallow_untyped_defs=True
disallow_incomplete_defs=True
check_untyped_defs=True
disallow_untyped_decorators=True

no_implicit_optional=True
strict_optional=True

warn_redundant_casts=True
warn_unused_ignores=True
warn_no_return=True
warn_return_any=False
warn_unreachable=True

strict_equality=True

ignore_missing_imports=False
```

```
./.venv/bin/mypy test
```

Comment (sbdchd) [reply hiredis/hiredis.pyi:1]: https://docs.python.org/3/c-api/arg.html was helpful for figuring out the arg types

File: hiredis/hiredis.pyi [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+class Reader:
+    def __init__(
+        self,
+        protocolError: Callable[[str], Exception] = ...,
+        replyError: Callable[[str], Exception] = ...,
+        encoding: Optional[str] = ...,
+        errors: Optional[str] = ...,
+    ) -> None: ...
+    def feed(
+        self, __buf: Union[str, bytes], __off: int = ..., __len: int = ...
```
Comment (sbdchd) [inline hiredis/hiredis.pyi:16]: we use `__` to signify that the params can't be passed in as keyword args

**Comment by ifduyue** [comment]
Very detailed comments. Good work!

**Pull request merged**